File: _138_copy_list_with_random_pointer/main.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (head == NULL)
        {
            return NULL;
        }

        unordered_map<Node*, Node*> associations;

        // First loop
        Node* newHead = new Node(head->val);
        newHead->random = head->random;
        associations.insert({head, newHead});
        Node* previous = newHead;
        Node* current = head->next;
        while (current != NULL)
        {
            Node* newNode = new Node(current->val);
            previous->next = newNode;
            newNode->random = current->random;

            associations.insert({current, newNode});

            previous = newNode;
            current = current->next;
        }

        // Second loop
        current = newHead;
        while (current != NULL)
        {
            current->random = associations[current->random];
            current = current->next;
        }

        return newHead;
    }
};

File: _138_copy_list_with_random_pointer/main.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None

        newHead = Node(head.val)
        newPrevious = newHead
        current = head.next

        correspondances = {}
        correspondances[head] = newHead

        while current:
            newNode = Node(current.val)
            newPrevious.next = newNode
            newPrevious = newNode
            correspondances[current] = newNode
            current = current.next

        newCurrent = newHead
        current = head

        while current:
            if current.random:
                newCurrent.random = correspondances[current.random]
            newCurrent = newCurrent.next
            current = current.next

        return newHead